Kernel+Userland: Remove assumption about pts and tty names in the Kernel

supercomputer7 opened the pull request:
This is done in a couple of aspects:
1. Remove MasterPTY::pts_name method and TTY::tty_name method. This
affect the Kernel in many places where we relied on these devices to
know where these are located at.
2. Remove ptsname and ttyname syscalls. These syscalls were an
abstraction layer violation, due to the fact they relied on assumed
paths to TTY and SlavePTY devices.
3. Provide support for ttyname(3) and ptsname(3) in LibC. Because we
can't use the kernel to determine these values for us, we rely
completely on LibC functions to determine these strings.
4. Create SelfTTYDevice with major number 5, minor number 0. We used to
create a symlink in SystemServer between /dev/tty to /proc/self/tty.
However, because /proc/self/tty relied on assumption of the TTY name,
it was removed. To supply support for operation on the current TTY of
the process, we create this device to replace the symlink at /dev/tty.
Internally, this device merely uses the current process TTY to perform
read(2), write(2) and ioctl(2).

-----

This is obviously not ready for merging as I need to split this patch into multiple commits. However, it's working perfectly fine and I'd like to get comments on this, especially the Userland part :)

supercomputer7 commented:
I splitted the gigantic commit into multiple commits. I still need to ensure every commit can be built on its own but I'm pretty sure that's the case. I'll check this tomorrow.

supercomputer7 commented:
This is ready for review now. I tested all commits one after one, they all build just fine. When merging this, the Kernel doesn't assume paths of any TTY or pseudo terminal device which is a very good thing. The only assumed paths (if I've not missed any other string) are the location of the `SystemServer` binary when we boot the OS, which is quite OK.

kleinesfilmroellchen commented on Kernel/TTY/TTY.cpp:
```diff
@@ -575,7 +570,8 @@ ErrorOr<void> TTY::ioctl(OpenFileDescription&, unsigned request, Userspace<void*
 ErrorOr<NonnullOwnPtr<KString>> TTY::pseudo_path(const OpenFileDescription&) const
 {
-    return tty_name().try_clone();
+    // return tty_name().try_clone();
```
Remove the comment?

supercomputer7 replied:
Oops, my bad

BertalanD commented on Userland/Libraries/LibC/stdlib.cpp:
```diff
@@ -696,6 +699,26 @@ int atoi(const char* str)
     return value;
 }
 
+// non standard function
+char* itoa(int value, char* buffer, int base)
```
glibc has `_itoa`, and musl has a static helper function for this exact operation. I'm not strongly against this, but we should check if we break some ports by defining a non-standard function in LibC.

supercomputer7 replied:
How can I check this? We don't lint ports (as the build part will take forever).

BertalanD commented on Userland/Libraries/LibC/stdlib.cpp:
```diff
@@ -719,8 +742,36 @@ char* ptsname(int fd)
+
+    // Note: glibc says that for a maximum size for `int' of 8 bytes we never need more than 20 digits.
+    // we create buffer to use to print the number in and it's 21 bytes to accomodate a null byte at the end.
+    char number_converted_string[21];
+    // Clean the array, just in case...
+    memset(&number_converted_string, 0, 21);
+    const char* devpts = "/dev/pts/";
+    const size_t devpts_length = strlen(devpts);
```
I think it would be cleaner to just use a `StringBuilder` and here, and just copy that to the destination buffer.

supercomputer7 replied:
I thought so too, but what about ports that compile without C++ support?

supercomputer7 replied:
So after you explained we link LibC against libstdc++ and therefore we can use C++ code, this seems like a good idea, and I will do that :)

BertalanD commented on Userland/Libraries/LibC/unistd.cpp:
```diff
@@ -398,11 +401,71 @@ ssize_t pwrite(int fd, const void* buf, size_t count, off_t offset)
+            size_t name_length = strlen(directory_name) + strlen(entry->d_name) + 1;
+
+            if (name_length > size) {
+                if (closedir(dirstream) < 0)
+                    return -1;
```
You can use a `ScopeGuard` here to avoid having to manually close the directory.

supercomputer7 replied:
So after you explained we link LibC against libstdc++ and therefore we can use C++ code, this seems like a good idea, and I will do that :)

BertalanD commented on Userland/Libraries/LibC/unistd.cpp:
```diff
@@ -398,11 +401,71 @@ ssize_t pwrite(int fd, const void* buf, size_t count, off_t offset)
+    struct dirent* entry;
+    char* name_path;
+
+    while ((entry = readdir(dirstream)) != nullptr) {
```
Not that we should care about it, but I think using `readdir` here breaks the reentrancy guarantees of `ttyname_r`.

supercomputer7 replied:
So after you explained we link LibC against libstdc++ and therefore we can use C++ code, this seems like a good idea, but probably not suitable for this PR. A FIXME will be added though to keep note of this.

supercomputer7 commented on Kernel/Devices/SelfTTYDevice.cpp:
```diff
@@ -0,0 +1,76 @@
+
+UNMAP_AFTER_INIT NonnullRefPtr<SelfTTYDevice> SelfTTYDevice::must_create()
+{
+    auto random_device_or_error = DeviceManagement::try_create_device<SelfTTYDevice>();
+    // FIXME: Find a way to propagate errors
+    VERIFY(!random_device_or_error.is_error());
+    return random_device_or_error.release_value();
```
Oops, my bad here.

supercomputer7 commented on Kernel/Devices/SelfTTYDevice.cpp:
```diff
@@ -0,0 +1,76 @@
+    RefPtr<TTY> tty = current_process.tty();
+    if (!tty)
+        return false;
+    return tty->can_read(open_file_description, offset);
```
Oops, my bad here

supercomputer7 commented on Kernel/Devices/SelfTTYDevice.cpp:
```diff
@@ -0,0 +1,76 @@
+{
+    auto& current_process = Process::current();
+    RefPtr<TTY> tty = current_process.tty();
+    if (!tty)
+        return Error::from_errno(ENODEV);
```
> Can we somehow prevent the user from opening /dev/tty if the process has no controlling TTY? Maybe return ENXIO from open?

Yes, and we probably should return the underlying TTY as @awesomekling suggested.

supercomputer7 commented:
I addressed the comments, leaving only the comment about ports and the `itoa` function to be solved now :)

supercomputer7 commented:
> I addressed the comments, leaving only the comment about ports and the `itoa` function to be solved now :)

The last comment was addressed too, so we don't need the `itoa` function anymore. Thanks @BertalanD for the awesome suggestions :)

supercomputer7 commented:
Rebased after merge conflicts (that were introduced after #12754 was merged).

awesomekling commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -442,7 +442,7 @@ ErrorOr<String> ptsname(int fd)
     auto* name = ::ptsname(fd);
     if (!name)
-        return Error::from_syscall("ptsname"sv, -errno);
+        return Error::from_string_literal("Could not find pts"sv);
```
We already have many other things using `Error::from_syscall()` that aren't *actually* kernel system calls. So I think it's better to leave this alone instead of making the error slightly worse.

awesomekling commented on Userland/Libraries/LibC/stdlib.cpp:
```diff
@@ -719,8 +722,45 @@ char* ptsname(int fd)
+    // FIXME: Use some sort of inline string (InlineString) from the AK library when it's ready!
+    // Note: glibc says that for a maximum size for `int' of 8 bytes we never need more than 20 digits.
+    // we create buffer to use to print the number in and it's 21 bytes to accomodate a null byte at the end.
+    char number_converted_string[21];
```
This array is not actually used for anything?

awesomekling commented on Userland/Libraries/LibC/unistd.cpp:
```diff
@@ -398,11 +402,72 @@ ssize_t pwrite(int fd, const void* buf, size_t count, off_t offset)
+        closedir(dirstream);
+    });
+
+    struct stat st;
+    struct dirent* entry;
+    char* name_path;
```
Please declare variables as late as possible. We are not bound by old C rules here :)
Also please initialize primitive variables explicitly instead of leaving them uninitialized.

supercomputer7 commented:
This is not stale...
